src/util/thread_metrics.rs

```diff
@@ -0,0 +1,217 @@
+                                                   "Total user and system CPU time spent in \
+                                                    seconds by threads.")
+                                             .namespace(namespace),
+                                         &["name"])
+            .unwrap();
+        let descs = cpu_totals.desc().into_iter().cloned().collect();
+
+        let pairs = try!(get_threads(pid));
+        let core = ThreadsCore {
+            refresh: 0,
+            thread_ids: pairs,
+            cpu_totals: cpu_totals,
+        };
+
+        Ok(ThreadsColletcor {
+            pid: pid,
+            descs: descs,
+            core: Mutex::new(core),
+        })
+    }
+}
+
+impl Collector for ThreadsColletcor {
+    fn desc(&self) -> Vec<&Desc> {
+        self.descs.iter().collect()
+    }
+
+    fn collect(&self) -> Vec<proto::MetricFamily> {
+        let mut core = self.core.lock().unwrap();
+        core.refresh += 1;
+        if core.refresh >= REFRESH_CNT {
+            core.thread_ids = get_threads(self.pid).unwrap();
+            core.refresh = 0;
+        }
+
+        for &(tid, ref tname) in &core.thread_ids {
+            if let Ok((utime, stime)) = find_thread_cpu_time(self.pid, tid) {
+                let total = (utime + stime) / *CLK_TCK;
+                let cpu_total = core.cpu_totals.get_metric_with_label_values(&[tname]).unwrap();
+                let past = cpu_total.get();
+                let delta = total - past;
+                if delta > 0.0 {
+                    cpu_total.inc_by(delta).unwrap();
+                }
+            }
+        }
+        core.cpu_totals.collect()
+    }
+}
+
+fn get_threads(pid: pid_t) -> Result<Vec<(pid_t, String)>> {
+    let mut pairs = Vec::new();
+
+    let dirs = try!(fs::read_dir(format!("/proc/{}/task", pid)));
+    for task in dirs {
+        let tid = match try!(task).file_name().to_str() {
+            Some(tid) => try!(tid.parse().map_err(|e| to_err(format!("{}", e)))),
+            None => return Err(to_err(format!("fail to read {} task", pid))),
+        };
+
+        let mut tname = String::new();
+        try!(fs::File::open(format!("/proc/{}/task/{}/comm", pid, tid))
+            .and_then(|mut f| f.read_to_string(&mut tname)));
+
+        // Pop tail '\n'.
+        tname.pop();
+        pairs.push((tid, sanitize_thread_name(tname)));
+    }
+
+    Ok(pairs)
+}
+
+static TH_CNT: AtomicUsize = ATOMIC_USIZE_INIT;
+
+fn sanitize_thread_name(name: String) -> String {
+    let mut san = String::with_capacity(name.len());
+    for c in name.chars() {
+        match c {
+            // Prometheus label characters `[a-zA-Z0-9_:]*`
+            'a'...'z' | 'A'...'Z' | '0'...'9' | '_' | ':' => {
+                san.push(c);
+            }
+            '-' => san.push('_'),
+            _ => (),
+        }
+    }
+
+    if san.is_empty() {
+        format!("unkonw_{}", TH_CNT.fetch_add(1, Ordering::Relaxed))
```

Comment:
the unknown thread may use different counter when call sanitize_thread_name multi times.